clients/drcachesim/tests/invariant_checker_test.cpp

```diff
@@ -686,6 +696,207 @@ check_function_markers()
+            gen_instr(TID, SYSCALL_PC, SYSCALL_SZ),
+            gen_marker(TID, TRACE_MARKER_TYPE_SYSCALL, 15),
+            gen_marker(TID, TRACE_MARKER_TYPE_TIMESTAMP, 16),
+            gen_marker(TID, TRACE_MARKER_TYPE_CPU_ID, 3),
+            // syscall xfer.
+            gen_marker(TID, TRACE_MARKER_TYPE_KERNEL_XFER, SYSCALL_PC + SYSCALL_SZ),
+            gen_marker(TID, TRACE_MARKER_TYPE_TIMESTAMP, 17),
+            gen_marker(TID, TRACE_MARKER_TYPE_CPU_ID, 3),
+
+            gen_marker(TID, TRACE_MARKER_TYPE_FUNC_ID, 2),
+            gen_marker(TID, TRACE_MARKER_TYPE_FUNC_RETADDR, CALL_PC + CALL_SZ),
+        };
+        if (!run_checker(memrefs, false))
+            return false;
+    }
+    // Incorrect: signal not arriving between a branch instruction and the function
+    // entry should not have a function ID marker after syscall xfer.
+    {
+        constexpr addr_t SIG_HANDLER_PC = 400;
+        constexpr addr_t SYSCALL_PC = 500;
+        constexpr size_t SYSCALL_SZ = 2;
+        constexpr size_t RETURN_SZ = 3;
+
+        std::vector<memref_t> memrefs = {
+            gen_instr(TID, 1),
+            gen_marker(TID, TRACE_MARKER_TYPE_KERNEL_EVENT, 1),
+
+            gen_instr(TID, SIG_HANDLER_PC),
+            // Return from the signal handler.
+            gen_instr_type(TRACE_TYPE_INSTR_RETURN, TID, SIG_HANDLER_PC + 1, RETURN_SZ),
+            // sigreturn.
+            gen_instr(TID, SYSCALL_PC, SYSCALL_SZ),
+            gen_marker(TID, TRACE_MARKER_TYPE_SYSCALL, 15),
+            gen_marker(TID, TRACE_MARKER_TYPE_TIMESTAMP, 16),
+            gen_marker(TID, TRACE_MARKER_TYPE_CPU_ID, 3),
+            // syscall xfer.
+            gen_marker(TID, TRACE_MARKER_TYPE_KERNEL_XFER, SYSCALL_PC + SYSCALL_SZ),
+            gen_marker(TID, TRACE_MARKER_TYPE_TIMESTAMP, 17),
+            gen_marker(TID, TRACE_MARKER_TYPE_CPU_ID, 3),
+            // There should not be a function ID marker here.
+            gen_marker(TID, TRACE_MARKER_TYPE_FUNC_ID, 2),
+        };
+        if (!run_checker(memrefs, true,
+                         { "Function marker should be after a branch", TID,
+                           /*ref_ordinal=*/12, /*last_timestamp=*/17,
+                           /*instrs_since_last_timestamp=*/0 },
+                         "Failed to catch function marker not after branch"))
+            return false;
+    }
+    // Correctly handle nested signals with the first one arriving between a branch
+    // instruction and the function entry.
+    {
+        constexpr addr_t BASE_PC = 100;
+        constexpr addr_t SIG1_PC = 200;
+        constexpr addr_t SIG2_PC = 300;
+        constexpr size_t INSTR_SZ = 1;
+        constexpr size_t RETURN_SZ = 3;
+        constexpr size_t SYSCALL_SZ = 2;
+
+        std::vector<memref_t> memrefs = {
+            gen_instr_type(TRACE_TYPE_INSTR_DIRECT_CALL, TID, BASE_PC, CALL_SZ),
+            // First signal.
+            gen_marker(TID, TRACE_MARKER_TYPE_KERNEL_EVENT, BASE_PC + CALL_SZ),
```

Comment:
Same comment: please make the call target==interruption PC something realistic.

Ditto for other cases below.

Reply:
I've updated the callee PC here and below.